### ISSUE TITLE
Tuple support for separated_list(), tablewriter, tablereader

### ISSUE THREAD

**Issue by JadeMatrix** [jtv/libpqxx#97]
First of all, I do see `tablewriter` and `tablereader` are marked as deprecated — is there currently a preferred alternative?  If there isn't one, what exactly are the multibyte issues these run into, so I can take a stab at fixing/replacing these classes.

I've found `tablewriter` to be pretty useful except for having to convert all values in a row into a single type.  The class is written such that it mentions "tuples" but doesn't actually have support for tuple types (multiple elements of different types).  I took a stab at implementing support for my own use, visible [in my fork of libpqxx](https://github.com/jtv/libpqxx/compare/master...JadeMatrix:inserter-tuples).  This of course necessitated tuple support for `separated_list()`, which is not deprecated.  If you want, I can make a pull request specifically for that.

I isolated a proof-of-concept implementation for `separated_list()` [in this gist](https://gist.github.com/JadeMatrix/1f01dacc650f233165a0d3eead3a5098); note this also supports plain arrays due to switching from `c.begin()`/`c.end()` to `begin(c)`/`end(c)` like C++11's range-based-`for` uses.

**Comment by jtv** [comment]
To be honest, I really just deprecated those classes because I felt I _didn't understand_ what multibyte issues there might be.

For instance, what if you've got an encoding where a multibyte character may contain a byte whose value happens to be equal to the ASCII value for backslash, tab, or newline?  It may sound like I made it up but once upon a time there was even an incompatible libpq change because this kind of thing was discovered to be a problem.  The tablereader code just ignores the whole issue.

It may not even an issue in practice nowadays, since it can't happen in UTF-8.  But not understanding something as basic as that makes me nervous!  Perhaps I should just say "no weird stuff please," like I did with arrays.

In the years after deprecation I was surprised by the number of people who found these classes useful, so I never quite brought myself to get rid of them.  They're in a sort of limbo, which isn't good either.  Maybe it's time to bring them back.

But yes, the way they're tied to a single type is a weakness.  The mention of `TUPLE` goes back to a day when the C++ standard library did not have tuples, and the word was used for database rows.  For a while the `row` class was called `tuple`.  These days, I do agree it would make sense to support proper tuples.  If that can be done without radical changes, then I'm definitely interested!

**Comment by jtv** [comment]
I have incorporated `std::begin`/`std::end` as you suggest.  I had completely forgotten about those.  And while I was at it I applied it wherever I could, just for uniformity.  Once I move the baseline to C++14 or beyond, I'll be able to replace the `rbegin` calls as well.

**Comment by JadeMatrix** [comment]
I've been in C++-mode recently, completely forgot that Postgres has another meaning for the word "tuple."

If I recall correctly they're not `std::begin()`/`std::end()`,  they're actually looked up as `begin()`/`end()` in the global namespace using ADL, so they can be defined for custom types.

I haven't looked at `tablereader` yet so I don't know how hard it would be to add tuple support would be there; I'll take a look later.  As far as `tablewriter` and `separated_list()`, I've got a working example over in my fork (I've non-exhaustively tested it in my code).

**Comment by jtv** [comment]
AIUI `begin` and `end` are indeed looked up using ADL, so that you can extend them for your own types even if they don't have matching member functions, but they're still in `std`.

**Comment by JadeMatrix** [comment]
Yeah I just re-read the description of range-based-for, the lookup is more complicated than just `begin`/`end` and `std::begin`/`std::end` essentially replicate its lookup so you're correct.

**Comment by JadeMatrix** [comment]
Well, I cleaned up the `tablewriter` tuple support and will whip up a test program next.  I took a look at `tablereader` and it seems pretty straightforward to do, except it's not clear what action should be performed when a row can't be converted into a tuple.  For containers, `tablereader::tokenize()` simply inserts all extracted fields so there's only an error state if no line could be read.

For tuples, the from-string specializations/overloads will handle their own errors so I'm not worried about that.  However, there could be a mismatch between the number of fields the `tablereader` was constructed with and the number of tuple elements, and this can only be known *after* the line is pulled.  It seems like this should throw a `failure`, but should it also set `m_done = true` like `tablereader::get_raw_line()` does?

**Comment by JadeMatrix** [comment]
In an attempt to support null values in tuples I've been looking through the class hierarchy for `tablewriter`; it looks like something similar was once supported or planned to be supported (haven't looked through the git history).  One of the constructors takes a `Null` string which it then passes to the `tablestream` constructor; the other `tablewriter` constructors pass an empty string.  `tablestream` then sets this string to its private member `m_null` which is never used.

It looks like this should be relatively easy to fix (the `Null` string just needs to be passed on to the parent transaction in `transaction_base::BeginCopyWrite()`); it looks like adding support for custom delimiters should be easy to add in at the same time.

**Comment by jtv** [comment]
I'm not sure about `m_done`, but it seems at least consistent with existing behaviour to set it.  That said, when this fails, things are broken and it should be fine to leave the connection in a broken state.  I wouldn't worry too much about it at this stage.

The problem with encoding is such that, unfortunately, it's not enough to let the client pass the delimiter it wants.  That's because depending on the encoding, the delimiter as a series of bytes may, through sheer coincidence, occur _inside_ the encoding of a different character.  So it's not a matter of finding the delimiter pattern in the bytes we get; it's a matter of making sure we parse the bytes according to the client encoding, so that we know exactly _when_ to look for the delimiter.

**Comment by JadeMatrix** [comment]
It doesn't seem Postgres itself cares much about this issue:

>`COPY TO` will terminate each row with a Unix-style newline (“\n”). Servers running on Microsoft Windows instead output carriage return/newline (“\r\n”), but only for `COPY` to a server file; for consistency across platforms, `COPY TO STDOUT` always sends “\n” regardless of server platform.

Since the doc also states that 1) the delimiter must be a single-byte character and 2) `COPY TO` will never emit octal/hex escapes, it makes me wonder how exactly it returns data in one of these encodings that may contain "\n" bytes.  I'll have to do some research on that.

So far the support I currently have added is for `tablewriter` where the slicing is done Postgres-side.

**Comment by jtv** [comment]
Thanks for diving into this.  The problem as I understand it would be in `tablereader`: we get a line from the server, and in that line, we encounter what we _think_ is the delimiter, or a newline.  But it could be just a byte _inside a multibyte character_ which happens to have the same value as the delimiter or as `'\n'`.  It would send us that character in the current client encoding, un-escaped, so the only "proper" way to tell is to decode that encoding.

This couldn't happen with UTF-8, where no multibyte character contains any byte value which could also be a character on its own.  But right now I'm not even sure it can't happen with UTF-16!

**Comment by KayEss** [comment]
It can't happen with UTF16 if you read the values two bytes at a time (i.e. as 16 bit values in the right endian mode), but if you do byte oriented processing then you will make mistakes. 0x0a0a is the code point for GURMUKHI LETTER UU 'ਊ'. There are many UTF16 code units that have a 0x0a in one of the bytes. Encodings like SHIFT-JIS need to be properly decoded to find the line endings.

For pgasio I think the problem won't exist because it exposes the network data chunking directly and [Postgres will send the rows one message at a time](https://www.postgresql.org/docs/9.3/static/protocol-flow.html#PROTOCOL-COPY). Maybe it is libpq that adds the `\n` at the row end and merges the network data?

**Comment by JadeMatrix** [comment]
You're right, UTF-16/32 need to be read as 16/31-bit "words" — luckily neither is a [supported encoding in Postgres](https://www.postgresql.org/docs/10/static/multibyte.html#CHARSET-TABLE); in fact all supported encodings/charsets are fixed- or non-fixed-width based on single bytes.

@jtv I haven't had time to test `COPY` behavior yet but if my first guess is correct Postgres will escape *only the delimiter or "\n" byte* even in multi-byte sequences, which would result in output that is not very human-readable but very easily machine parsable.  This is again assuming the documentation is correct *and* `COPY` is safe to use with non-ASCII-compatible encodings.

@KayEss Your comment inspired me to look into this and apparently there *might* be a bug in libpqxx regarding this.  The call stack is a bit longer, but basically `tablereader` calls `PQgetCopyData()` per-line.  `PQgetCopyData()` on success allocates a line buffer, sets it to the `char**` passed in, and returns the byte size of the line.  However, libpqxx on `connection_base.cxx:958` simply sets the `std::string& Line` to the buffer pointer value, copying only up until the first null byte:

```cpp
  switch (PQgetCopyData(m_conn, &Buf, false))
...
  default:
    if (Buf)
    {
      std::unique_ptr<char, void (*)(char *)> PQA(
          Buf, freepqmem_templated<char>);
      Line = Buf;
    }
    Result = true;
  }
```

This should be

```cpp
  auto line_len = PQgetCopyData(m_conn, &Buf, false);
  switch (line_len)
...
      Line = {Buf, line_len};
```

`pqxx::tablereader::extract_field()` should then simply iterate over the line until it reaches the end of the `std::string` and not have an early-exit for "\n"; I do see a comment about backwards-compatibility here, so the this should probably instead discard the "\n" *only* if it is the last char in the string.

It's late now so I'll continue researching this week.

**Comment by jtv** [comment]
See?  I knew there'd be problems in the tablestream classes if we looked close enough.  :-)

For now I've incorporated your fix for nul bytes.  Thanks!

**Comment by JadeMatrix** [comment]
I'm not so sure it's a bug or not, because I don't think `COPY` would actually emit null bytes, but it could cause more than the required bytes to be copied to `Line` e.g. a sort of buffer overflow.  Maybe.  Either way fixing it is safer.

I'm trying to create a test for the `WIN866` encoding since it has characters that overlap with ASCII "\n"/"\t", but it seems Postgres has some issues with such an obscure encoding:

```
psql (10.3)
Type "help" for help.

postgres=# CREATE DATABASE copy_encoding_test
postgres-#     TEMPLATE template0
postgres-#     OWNER postgres
postgres-#     ENCODING 'WIN866'
postgres-# ;
ERROR:  encoding "WIN866" does not match locale "en_US.UTF-8"
DETAIL:  The chosen LC_CTYPE setting requires encoding "UTF8".
postgres=# CREATE DATABASE copy_encoding_test
postgres-#     TEMPLATE template0
postgres-#     OWNER postgres
postgres-#     LC_COLLATE 'ru_RU.cp866'
postgres-# ;
ERROR:  encoding "UTF8" does not match locale "ru_RU.cp866"
DETAIL:  The chosen LC_COLLATE setting requires encoding "WIN866".
postgres=# CREATE DATABASE copy_encoding_test
postgres-#     TEMPLATE template0
postgres-#     OWNER postgres
postgres-#     ENCODING 'WIN866'
postgres-#     LC_COLLATE 'ru_RU.cp866'
postgres-# ;
ERROR:  encoding "WIN866" does not match locale "en_US.UTF-8"
DETAIL:  The chosen LC_CTYPE setting requires encoding "UTF8".
postgres=#
```

😅
Found something about this on [a Russian BBS](http://www.sql.ru/forum/896329/enconing-win866-na-linux) and they mentioned using `localedef` but it seems my OS doesn't have support for the `IBM866` charmap.

**Comment by JadeMatrix** [comment]
With some further research into `WIN866`/`CP866` I've read that the code pages with dingbats in the ASCII control spots were [a one-time DOS thing](ftp://kermit.columbia.edu/kermit/charsets/cyrillic-summary.txt):

> For the PC community, Code Page 866 seems to be quite important. This is
> what Microsoft is using in its russified version of MS-DOS. In 0-31
> ascii control chars are replaced by a random selection of dingbats. In
> 32-126 it is identical to ascii, and in 127 it has something that looks
> like a little house (the interpretation of such positions seems to be
> subject to much uncertainty).

Taking a look at Python's `cp866.py` that handles encoding/decoding Code Page 866 files, it seems they only support the version with control codes in `0x00`-`0x1F` spots.

I've looked through the multibyte encodings supported by Postgres and still can't find one that conflicts with ASCII control codes.  I don't think I overlooked any, but I can put together a spreasheet to make sure.  However some of them do conflict with other characters, for example using `@` as a delimiter with the [`GBK`](https://en.wikipedia.org/wiki/GBK_(character_encoding)#Encoding) encoding:
<img width="1440" alt="screen shot 2018-06-19 at 12 48 12 am" src="https://user-images.githubusercontent.com/1753533/41577375-8112b928-735a-11e8-8c3a-2f3d1ced0a87.png">

Conclusion: `tablereader`/`tablewriter` *should* be safe to use as long as an ASCII control character is used as the delimiter.  @jtv I saw you incorporated the line-size fix; I'll have to do some testing to see if it's safe to discard trailing "\n"s; if it is, there should be no issues using these two classes for all supported encodings as long as a control-code delimiter is chosen.  (Luckily "\t" is already the pqxx default.)

**Comment by JadeMatrix** [comment]
Addendum — Looks like the safest bet is to have pqxx handle this internally using "\t" as the delimiter and the default null string.  Postgres *will* escape the delimiter with a backslash when it appears *outside* of a multi-byte sequence.  The problem is nulls; right now `tablereader`/`tablewriter` don't support nullable types, but adding tuple support will require this.  Using an empty string conflicts with real empty strings, and the Postgres default of `\N` (two bytes: backslash-"N") means libpqxx would need to do the encoding-specific escaping we're trying to avoid as the byte `\` can appear in multi-byte characters.  Luckily Postgres has a handful of escape sequences that `COPY` can emit.

I propose using something like "\f" (form feed) for nulls internally to `tablereader`/`tablewriter` as Postgres will escape these as "\f" in strings but use the literal in nulls:

<img width="939" alt="screen shot 2018-06-19 at 1 31 03 am" src="https://user-images.githubusercontent.com/1753533/41578467-82c5c0d4-7360-11e8-95b4-4e31346e9e05.png">

**Comment by jtv** [comment]
Time for a quick anecdote.  Thanks @KayEss for digging up that name SHIFT-JIS.  I knew there was a complex Japanese encoding out there with this problem, but the name had escaped my mind.

Years ago, long enough that I forget most of the details, there was an _incompatible security patch_ to libpq.  They actually made an incompatible API change in a patch upgrade, even for previous releases.  You can imagine how big a deal that was!  They wouldn't have done it unless it was incredibly important.

Some function dealing with SQL strings now needed to know the encoding, as an added parameter.  The reason was that strings in SHIFT-JIS, and possibly other encodings, might contain bytes that would look like quotes or backslashes.

And that's how I know that this problem exists.  :-)  I do hope that such encodings have been slipping away into oblivion, but out in the real world, we see some hopeless old software still keeping the engines running.

**Comment by JadeMatrix** [comment]
Well, this won't help with interpreting escape sequences in strings I'm afraid; there's not much you can do about that other than linking with a library with full support for Postgres' encodings.  This just solves the delimiter/null-sequence ambiguity for splitting lines.  Parsing individual fields is still left up to `pqxx::string_traits` specializations.

**Comment by jtv** [comment]
What bugs me is that we don't even really need full support for the encodings — we just need to know how many bytes to skip next when scanning for a pattern!

Maybe all we need is a mapping  from postgres encoding codes to skipping functions, and an error for encodings that aren't supported.  In most cases there's nothing to skip — even in UTF-8 there's really no reason to take on the extra complexity.

**Comment by JadeMatrix** [comment]
Actually yeah that's a good point — most multi-byte encodings are fairly well organized in that way, even `GBK` and `GB18030`.  I can throw together a proof-of-concept when I go through the Postgres encodings list again.

**Comment by jtv** [comment]
That wou

**Comment by JadeMatrix** [comment]
Quick update — I've got a test program that in theory will step through strings encoded in Posgtres' supported encodings, which was pretty easy.  Now I'm working on generating test data.  I've taken several approaches but unfortunately Postgres and Python disagree on what characters are valid for some of the encodings.  Quite a pain but I'll figure out something.

**Comment by jtv** [comment]
IIRC Python uses UTF-16 internally...  Which means you get the whole
headache of surrogate pairs.  But postgres has also gone through similar
pains in the past as checking of valid Unicode characters became stricter.
Still better than the mess of
these-bytes-are-encoded-but-we-don't-really-know-how that databases used to
have though!


Jeroen

**Comment by JadeMatrix** [comment]
Well I was trying to use Python's built-in `codecs` library to filter the Unicode character set for insertion into test tables (for later `COPY` & parsing).  But Python's `BIG5`, `EUC_JP`, and `SHIFT_JIS_2004` differ slightly (not a problem), `EUC_JIS_2004` and `EUC_KR` differ majorly, and it doesn't even support `MULE_INTERNAL` (understandably) or `EUC_TW`.

What I'll do next is instead iterate through each character in Unicode & try insertion into a table with the desired encoding set for that column, skipping any insertion failures.  That'll let Postgres handle the encoding.

**Comment by jtv** [comment]
Any news?  :-)

**Comment by JadeMatrix** [comment]
Yeah, almost done with testing, but Postgres doesn't support conversion between `UTF8` and `MULE_INTERNAL`, so I'm trying to figure out how to even generate a table to `COPY`.  All the other encodings are easy, simply iterate through `UTF8`, try to convert to the target encoding using `convert_to()`/`convert_from()` and skip it if the conversion fails.

**Comment by JadeMatrix** [comment]
Well, some good news and some bad news.  Good news is I've got code that iterates through all but one supported encoding (and gives useful error messages, too!).  Also good news is I was able to write a PLPGSQL script to generate the entire `MULE_INTERNAL` encoding — it only took 18 and a half hours to run.

Bad news is apparently the `MULE_INTERNAL` encoding Postgres uses doesn't match up with the [description here](https://www.xemacs.org/Documentation/21.5/html/internals_26.html#Internal-Mule-Encodings) — I even tried to implement the basic version that only checks for lead-vs-trail bytes.  That is, according to the doc:

>The basic characteristic of this encoding is that the first byte of all characters is in the range 0x00 - 0x9F, and the second and following bytes of all characters is in the range 0xA0 - 0xFF. This means that it is impossible to get out of sync […]

However, Postgres spat out a CSV (via `COPY`) that contained a single-byte sequence `0xA0` as its 133rd row (my test files are one character per row).  The only options left are to dive into the Postgres source to see what they're actually doing, or just not support `MULE_INTERNAL`.  Honestly not supporting it probably wouldn't be that bad a choice in my opinion, Postgres already doesn't play well with the encoding unless *everything* is done in that encoding (even the database must be created with it).

Anyways I'll try to track down the Postgres source for the encoding and see if it's readable.

**Comment by jtv** [comment]
Would this be a good issue to punt to the core postgres team?  I don't know anything about this encoding... might turn out they shouldn't be claiming to support it either.

If we have a way of scanning for character boundaries in all postgres-supported encodings except `MULE_INTERNAL` (which by the way I wasn't even aware of), then we can try building it into the tablereader class and the array parser.

**Comment by JadeMatrix** [comment]
I copied over & reformatted the iteration code I came up with in [this commit](https://github.com/JadeMatrix/libpqxx/commit/2bd8701816f0960e091f790292ffcadbdc9d1b6c).  It compiles under CMake, but none of the other code uses it yet.  I'd appreciate it if you took a quick look over the two source files to make sure the code style matches and that the interface is to your liking (esp. use of exceptions for encoding errors — hopefully unlikely as this'll be used for stuff coming from Postgres).

Next I'll make sure it builds with your other build options, then start figuring out how to integrate the encoding stuff where needed.

**Comment by jtv** [comment]
Okay, skipping that line and moving on to the copyright line, which also needs an update.  :-)  We can argue about whether my name should be in there, but the year 2001 definitely shouldn't!

**Comment by jtv** [comment]
That's a pretty long enum for the encodings!  I have all sorts of questions about it...
 * Is there any relation to the values returned by `PQclientEncoding`?
 * Should there be?
 * If not, since we only care about character _boundaries_ here, wouldn't it make more sense to lump all single-byte encodings into one enum value?
 * Better not have a commented-out value in the middle...  Uncommenting it would change the numeric values for all the entries below it.

**Comment by jtv** [comment]
Why does `THROW_FOR_ENCODING_ERROR` need to be a macro?

**Comment by jtv** [comment]
I guess `seq_position` does a job that could also be done by `std::string_view` — which unfortunately wasn't added to the standard until C++17.

But I wonder if there isn't a simpler problem hiding underneath: "here's a pointer to my current character — where does it end?"

**Comment by JadeMatrix** [comment]
The throw macro doesn't need to be one, I've changed it to a `[[noreturn]]` function.

I looked into using the result of `PQclientEncoding()`, unfortunately libpq explicitly warns against using the enum for these values:

```
 *	NOTES
 *		This is used both by the backend and by libpq, but should not be
 *		included by libpq client programs.  In particular, a libpq client
 *		should not assume that the encoding IDs used by the version of libpq
 *		it's linked to match up with the IDs declared here.
```

The "official" way to switch on encodings would be to call `pg_encoding_to_char()` to get the string name of the encoding.  So that we're not doing a string match for every lookup, we'd want to convert this to some libpqxx-internal enum value, which leaves us back where we started.  So I suppose it could be an enum of encoding classes: `MONOBYTE`, `EUC_JPLIKE`, `UTF8`, etc. for each of the iterate functions.  I'll add functions for converting to the enum from libpq encoding `int`s and string names.

**Comment by jtv** [comment]
This gets better and better.  :-)  Thanks again for doing all this work.

Also, a purely pedantic thought: would it make sense to _templatise_ the per-encoding functions on the value of that enum?  That might open the door to some optimisations, with lots of inlining and such.

**Comment by JadeMatrix** [comment]
So on second thought I would only combine single-byte encodings into `MONOBYTE`, as those can't have iteration errors.  The rest I would leave separate because the error message is `invalid sequence ... for encoding NAME`, unless you want to rewrite this so exceptions are thrown higher up where the encoding name can otherwise be known.

Templatizing the per-encoding functions does make sense; I could also leave the single-byte implementation as the default.  It wouldn't make sense to have the `next_seq()` function itself be a template, because the connection encoding is only known at runtime, so that would have to keep the switch.  Would you like the per-encoding template functions exposed in the header (just the signatures) or kept in a private namespace in the `.CXX`?

**Comment by jtv** [comment]
Well as you say, the caller will need that runtime switch.  But I'm hoping that maybe some loops inside libpqxx can be templated, so that the runtime switch is only needed _outside_ the loop.  Might buy us some speedup.  :-)

**Comment by JadeMatrix** [comment]
That's kind of what I figured you meant.

On a side note, the libpq/Postgres header that contains the encodings enum also has a description of the `MULE_INTERNAL` encoding and how it differs from both the emacs and Xemacs versions.

**Comment by JadeMatrix** [comment]
Okay, I've got something like this right now:

```cpp
// Runtime version, implemented with a `switch()` in the .CXX file
seq_position next_seq(
  encoding_group enc,
  const char* buffer,
  std::string::size_type buffer_len,
  std::string::size_type start
);

// Template version
template<encoding_group E = encoding_group::MONOBYTE>
seq_position next_seq(
  const char* buffer,
  std::string::size_type buffer_len,
  std::string::size_type start
)
{
  // Monobyte implementation
}

// Template specializations
// As such, these will be actual, compiled functions so must either be marked
// `inline` or be implemented in the .CXX file
template<>
seq_position next_seq</* encoding group enum value */>(
  const char* buffer,
  std::string::size_type buffer_len,
  std::string::size_type start
)
{
  // Implementation for encoding group
}
```

I'm not sure templatizing these will do much, as template specializations are real, compiled functions at basic optimization levels and would be built into the library.  I haven't tested it, but I assume they would also need to be marked `inline` or implemented in the `.CXX` in order to avoid linking errors within libpqxx or with applications using libpqxx.  So making these functions templates doesn't do much but change the syntax by which they're called.

**Comment by jtv** [comment]
You can just put the specialisations in the ".cxx" file, non-inline.  Bear in mind there's no generic implementation, only specific specialisations for the encoding groups.  A generic implementation would have had to be inline.

Better not have a default value for that template parameter I think — makes it too easy to forget that an encoding group _must_ be specified.

**Comment by jtv** [comment]
Oh by the way, if it simplifies the signatures, don't be afraid to speak entirely in pointers: begin, end.  No need to pass full knowledge about the buffer as a whole.  It seems cleaner, but in practice it just tempts readers to think that the functions underneath have some reason to know where the buffer starts.  All that _really_ matters is where the character starts, where the buffer ends, and where the character ends!

**Comment by JadeMatrix** [comment]
I'll go ahead and remove the default template parameter, that's a good point.

I actually had a version that only took an offset pointer and a remaining count, but that doesn't let you generate errors that say "Invalid sequence ... **at byte X**".  I found out the hard way that's indispensable for debugging encoding errors.  This is only *intended* to be used internally, so I feel a more complex signature is a small price to pay for better error messages.  That is, unless you want to push up error generation to a higher level, which means more information will have to be contained in the return type.  Right now there's two possible groups of return values:

* `{start, end}` for valid sequences
* `{std::string::npos, undefined}` for end-of-string

The end-of-string value was to make *using* these functions simpler, because you only had to check the return value and not your current position in the buffer.  If error generation should be handled higher up, it probably makes more sense for that to be the encoding-error state and let the caller keep track of the buffer position.  However, you lose all information about the number of offending bytes, so you can't generate PostgreSQL-esque error messages.  If this required a minimum of C++17 `std::variant` or `std::optional` would be the obvious choice.

**Comment by JadeMatrix** [comment]
Here's the latest diff: https://github.com/jtv/libpqxx/compare/8363798...JadeMatrix:inserter-tuples

It's a diff from a specific commit because I have some `tablewriter`/`tablereader` tuple support committed before the encoding stuff on this branch.  Once this part is done you can decide how you want the commits merged.  I can either complete tuple support first, or rebase to a fresh branch (preserving commit dates, too) that can be merged immediately.

**Comment by JadeMatrix** [comment]
Working on adding this to `tablereader` right now, and there's a couple things I'd like to discuss regarding the implementation:
* Should `tablewriter`/`tablereader` still support custom null/delim strings?  I already had it implemented before starting on the encoding stuff, just wondering if you feel it's worth keeping that.
* Should `tablewriter`/`tablereader` be templated on the encoding group?  They can grab the encoding information from `transaction_base`.
  * If so, a `std::logic_error` (or variant) should be thrown somewhere if the template encoding and the connection encoding don't match.
  * If not, should the encoding info be grabbed at construction, for each row, or what?  Maybe just at construction and document "don't change encodings while table streaming," which should really be common sense anyways.
* I think `std::logic_error` (or a libpqxx variant) is the most appropriate exception to throw for bad table reads/writes, where the tuple element count doesn't match up with the number of columns.  Anything better to use?  I'm already thinking of changing encoding errors from `std::runtime_error` to `std::logic_error`.
* Should the function to get the client encoding group be a `public`/`protected` member of `transaction_base` or a free function taking a `transaction_base&`?  The former will require `pqxx/internal/encodings.hxx` to be included in `pqxx/transaction_base.hxx`, the latter the other way around.
* I added several variants of the function `find_with_encoding()` as an encoding-safe `std::string::find()` replacement.  Should I add `rfind_*` variants, too?  I don't think any libpqxx code will need them right now — `tablereader` only uses forward-find to locate the next tab character.

**Comment by jtv** [comment]
_Finally_ found a good time to go over your diff.  Sorry it's taking so long.  Will get into the discussion points you raise next.

And by the way I'd like to compliment you on the quality of what you're doing — you have good arguments and present them well.  As you say, good error messages are massively helpful when debugging problems.  The code looks clean and clear, which to me is perhaps even more important than correctness.  (It may sound ridiculous when I say that, but try maintaining unclear code for a decade, regardless of how correct it is!)  Your comments also give me a lot of confidence when reading the code.

I do have a few superficial nitpicks with the encoding error message, nothing dramatic:
1. Instead of "invalid sequence" I'd say something like "Invalid byte sequence for encoding X."   "Invalid sequence" is exactly right within context, but the user doesn't necessarily know _a priori_ what that context is.
2. I prefer capitalised, punctuated error messages nowadays.  :-)
3. There's a lot of information in that sentence.  May be a bit overwhelming.  Given punctuation as a feature, I think a few short sentences would be better than one long one.  The actual byte sequence should be an afterthought IMHO — for those who choose to try and debug the actual data.

The `find_with_encoding` is indeed exactly the sort of thing I was hoping for with the templatisation.  Although I may have been skirting the edges of Knuth's Law there.  :-)

I do still feel that it should probably be up to the loops calling `next_seq` to remember the starting point for any given call, but we can always see what works best later.  In all honesty I probably won't resist the temptation to mess with it and make changes that could be more to do with my personal taste than any objective improvements.

Oh, and something I've been wondering about: do we really need `static_cast` to turn `char` into `unsigned char`?  I'm never quite sure.

`DISPATCH_ENCODING_OPERATION` looks a bit redundant...  Did that become unnecessary with the templatisation?  I may be missing something, but the only thing I see it doing is "ignore this function call if the encoding group is not one of these."

Anyway, it's all small fry.  I'm quite happy with the encoding code as it stands so I'd like to get it integrated soon, and start writing unit tests for it.  That last bit is not a fun job, with lots of custom testing macros, so it's probably best if I take care of that part.

**Comment by jtv** [comment]
Next, your discussion points:

>      Should tablewriter/tablereader still support custom null/delim strings? I already had it implemented
>  before starting on the encoding stuff, just wondering if you feel it's worth keeping that.

Hmm...  now that you mention it, it does sound like unnecessary baggage.  My only doubt is: is encoding a factor?  I assume that the null string should be specified in the applicable client encoding.  Would it ever come out as a different byte string in any of the supported encodings?

>     Should tablewriter/tablereader be templated on the encoding group? They can grab the encoding
>  information from transaction_base.

Could make for a nice optimisation, but I wonder if it wouldn't just tempt people to hard-code bad assumptions about their client encodings.  My gut tells me to keep this a runtime thing, and then see about optimisations later.  The templates you already have take care of the inner loops, and those are normally the top priority.

>         If not, should the encoding info be grabbed at construction, for each row, or what? Maybe just at
>  construction and document "don't change encodings while table streaming," which should really be
>  common sense anyways.

To be honest, I don't even know that you _could_ change encodings while streaming!  I'd definitely go with the constructor option.  If it turns out to be a big, desirable feature (which I doubt) then I think it's fine to make it less pretty and convenient than the normal case.

>     I think std::logic_error (or a libpqxx variant) is the most appropriate exception to throw for bad
>  table reads/writes, where the tuple element count doesn't match up with the number of columns.
>  Anything better to use? I'm already thinking of changing encoding errors from std::runtime_error to
>  std::logic_error.

For the wrong-number-of-columns mistakes, I agree `std::logic_error` is a very fitting error.  And as you surmise there is a libpqxx variant: `pqxx::usage_error`.

The encoding errors are different AFAICS — that's bad data, which could come from anywhere.  I would suggest `pqxx::argument_error`.

>     Should the function to get the client encoding group be a public/protected member of
>  transaction_base or a free function taking a transaction_base&? The former will require
>  pqxx/internal/encodings.hxx to be included in pqxx/transaction_base.hxx, the latter the other way
>  around.

My preference would be to make it a member.  I use a system of "call gates" to specify very specific private access across classes in the library.  To call a private member function on an object of a different class, construct a "call gate" as a wrapper around the object, and it has "friend" access to the object.

Using these call gates, I tend to put these functions in the connection class as private members, and then have forwarding calls in the transaction class.

Also, including the header may not be required.  If there's no noticeable performance impact, I'd use forward declarations in the connection header, and only include the full encodings header from `connection_base.cxx`.

>     I added several variants of the function find_with_encoding() as an encoding-safe std::string::find()
>  replacement. Should I add rfind_* variants, too? I don't think any libpqxx code will need them right
>  now — tablereader only uses forward-find to locate the next tab character.

Let's not go _looking_ for trouble.  :-)

**Comment by JadeMatrix** [comment]
Regarding error messages, it seems I may have been mixing up several styles of encoding errors messages:

```
psql:
    ERROR:  invalid byte sequence for encoding "UTF8": 0xff
Postgres COPY:
    ERROR:  character with byte sequence 0xc4 0x80 in encoding "UTF8" has no equivalent in encoding "LATIN1"
Python:
    UnicodeEncodeError: 'ascii' codec can't encode character '\ua000' in position 0: ordinal not in range(128)
```

I'll go ahead and stick the bytes after a colon.

To get the appropriate overload for `std::ostream operator<<` to print the hex value of a `char`, a cast to `unsigned char` then `unsigned int` (or `int` I guess) is needed.  Just casting to `unsigned char` still inserts the ASCII character (ignoring the `std::hex` directive because "it isn't a number"), and simply casting to `unsigned int` first casts to signed `int`, so you get a whole bunch of leading `ff`s due to two's complement.

`DISPATCH_ENCODING_OPERATION` is a macro to simplify the implementation of the runtime-switched versions of the functions.  The template versions can just carry down the template argument, but you can't create templates at runtime so you have to manually specify each template specialization to use for each case.  And it won't ignore any other values — there can't be any other values because `pqxx::internal::encoding_group` is a C++11 `enum class` rather than C-style `enum`, and if any enum values are added in the future these switches will cause a compiler warning about unhandled cases (which is why in general you should be very careful about mixing `enum class` and `default:` cases).  The macro is so less code has to be copy-paste-edited, which reduces the possibility of accidental typos.

The only thing from `encodings.hxx` that `transaction_base` would need is `encoding_group`, and it couldn't be forward-declared because the function returns that type by value.  I'll separate the `enum class` into its own header which both can then include.

**Comment by jtv** [comment]
Excellent.

**Comment by JadeMatrix** [comment]
Regarding custom null/delim strings, it's been a while since I researched it, but if I remember correctly they are used as-is, byte-for-byte by Postgres.  The delimiter can only be a single one-byte character, so not many encoding issues there.  There's more flexibility with the null string, but in the default format `text` Postgres won't escape custom null strings that appear in data; if the null string is `NULL` and you have a text field with the value `'NULL'`, there's no way to tell the difference.  However, if the `COPY` specifies `FORMAT csv`, ambiguous strings are quoted rather than escaped, which disambiguates null strings.

For example, in `FORMAT text` with null string `NULL` and default delimiter `,`:

* `','::TEXT` becomes `\,`
* `NULL` becomes `NULL`
* `'NULL'::TEXT` becomes `NULL`

With the same options but `FORMAT csv`:

* `','::TEXT` becomes `","`
* `NULL` becomes `NULL`
* `'NULL'::TEXT` becomes `"NULL"`

Because dealing with quoted strings would involve a bit more work, I'd say drop custom delimiters and stick with handling splitting behind-the-scenes.  The existing `tablereader` reports null values as the string `\N`, but the tuple implementation can use `pqxx::string_traits<>::null()` when it sees a null string.

**Comment by jtv** [comment]
It's a nasty one, isn't it?  A string value that _may_ mean null.  :-(  All we can do is punt that one to the caller, with the current API.  But might it be worth considering a different API based on the CSV format?

As for the delimiter, yes, if there are no multi-byte issues with the custom delimiter, then let's definitely drop it.  Maybe someday somebody will hit a case where it's slow because too many characters look just like the delimiter, but we'll worry about optimising that then.  With any luck there'll be a better format for doing this before that day.

**Comment by JadeMatrix** [comment]
I left out one bit — in `FORMAT text`, `'\\N'::TEXT` becomes `\\N`, so no worries there.  It's only *custom* null strings that have the issue.

**Comment by jtv** [comment]
Phew.  :-)  No need for any customisation then, really?

**Comment by JadeMatrix** [comment]
Correct.

I'm working on finishing up the tuple support right now, and realizing there's quite a bit of awkwardness in the design of the string conversion utilities, at least the ones `tablewriter` is concerned with.  As the separated list is created, the `Escaper` accessor converts each to a string with `pqxx::to_string()`, checks that against a "null" string value, then passes either `"\N"` or the string back into `pqxx::to_string()`.  The double stringify is unfortunate but would need some far-reaching changes to fix.  The bigger issue is the check against a "null" string value; the user has to pick a value they're sure will never appear in a string they want to insert, which isn't very practical.  With the tuple version this isn't needed, as nulls can simply be checked for using `pqxx::string_traits<>::is_null()`.  However this would need a second escaper class that doesn't require a "null" string to check against.  I'd rather not have two separate classes, but retaining the container inserts is important for inserting into tables whose column count isn't known at compile-time, similar to the issue #75 pointed out.  (All the code in my personal projects knows ahead of time the table it's writing to, but I can't speak for everyone.)

However the fix for #75 also used a same-type container.  Perhaps a proper solution would be a user-creatable row type to which fields are appended (allowing a template `append<>()` member that will do proper stringification)?  I'll leave decisions about further design changes to you; I'd be glad to work on it, but it's a bit out-of-scope here.  For now I'll add a second type of escaper accessor for the tuple inserts that never checks against a null string, as that serves this ticket's goal.

**Comment by JadeMatrix** [comment]
Looks like I might have something that works.  I'll write some more thorough tests for it later this week, something suitable for including in libpqxx's test suite.

<img width="1440" alt="screen shot 2018-08-20 at 9 46 15 pm" src="https://user-images.githubusercontent.com/1753533/44375653-c1648080-a4c2-11e8-852f-c63c4c008a94.png">

**Comment by jtv** [comment]
That looks quite nice.  Might be worth it to support `std::optional` in compilers that have it — that would eliminate a lot of the problem cases for nulls.

**Comment by JadeMatrix** [comment]
I'm working on that at the moment, it's actually quite difficult as `std::optional` doesn't have much in the way to identify it signature-wise.  I've wound up with a multi-layer approach — types that are and aren't "dereferencable" (have an `operator *()`) are extracted differently.  Those that aren't are set to `from_string()` or `string_traits<>::null()`; those that are are set to `from_string()` or either `nullptr` or `std::nullopt`.

That last bit I haven't worked out yet; for compilers without `std::optional`/`std::experimental::optional`, I'll need a "null value" function that only ever returns `nullptr`, but for those that do it'll have to detect if there's an `operator =()` that takes `std::nullopt_t` I guess.  Any custom "optional" types that take either `nullptr` or `std::nullopt_t` cannot be used that way though, so I might have to figure out something that will only detect `std::optional`.  Perhaps something like 

```cpp
template<template<typename> O, typename = void> struct is_std_optional : std::false_type {};
template<template<typename> O> is_std_optional<std::enable_if<
    std::is_same<O<std::nullptr_t>, std::optional<std::nullptr_t>>::value
>::type> : std::true_type {};
```

I'll see if that even compiles.

**Comment by JadeMatrix** [comment]
Here's the latest (full) diff: https://github.com/JadeMatrix/libpqxx/compare/jtv:master...inserter-tuples

I "finished" a draft of `tablereader`, but I'm getting `no COPY in progress` errors after reading the first row.  I can't figure out why, because I haven't touched anything that should have broken that.  I'll have to do further investigation, might be an environment error.

I added a header file containing template metaprogramming utilities for dealing with stuff like `std::optional`.  This was mainly for implementing null-value support for `tablereader`, but it should be useful elsewhere in libpqxx too.  Actually, I should use them to make sure `tablewriter` can also handle `std::optional` etc.; I can do that next.

(Just realized the new `pqxx::internal::is_derefable<>` utility type doesn't work for types that aren't default-constructable, but I can't remember the right trick to use for that at the moment.)

**Comment by jtv** [comment]
I must admit I'm a little apprehensive to hear about this check for the dereference operator...  Can't it be done the way I did in `include/pqxx/internal/statement_parameters.hxx`?  If that means having additional functions where return values become reference parameters, so be it.

**Comment by JadeMatrix** [comment]
Doing it this way allows using more than `std::optional`, such as custom optional types or even pointers:

```cpp
int my_int;
std::string my_string;
auto got_values = std::make_tuple(
    &my_int,
    &my_string
);

tablereader >> got_values;

if (std::get<0>(got_values))
    std::cout << my_int << std::endl;
else
    std::cout << "null" << std::endl;
if (std::get<1>(got_values))
    std::cout << my_string << std::endl;
else
    std::cout << "null" << std::endl;
```

It may with some tweaking allow for `boost::optional`, which uses `boost::none` as its null-value (not `nullptr` or `std::nullopt`), possibly a "null traits" type users can specialize.

**Comment by jtv** [comment]
Time to add a `std::optional` specialisation template for `string_traits`?  Basically `string_traits<optional<T>>` would say a value is null if it's empty, and otherwise it can just defer to `string_traits<T>`.

(Last time I checked I still found C++17 support lacking in deployed systems, so I'll let the baseline C++ version hang back a few years, but might as well look ahead.)

**Comment by JadeMatrix** [comment]
That's a good point, a generic optional-string-traits is possible with this and would be ideal anyways.  I'll go ahead and work on that, it'll also reduce the `tablereader`/`tablewriter` tuple code size a bit.

`std::nullptr_t` will still have to have special treatment as far as to- and from-string goes, because nulls have to be detected *before* escaping/unescaping strings.

**Comment by JadeMatrix** [comment]
I've been working on spreading around the more generic optional detection, and I discovered the unit tests aren't building with CMake.  The CMake script for `unit_runner` globs its sources into `UNIT_TEST_SOURCES` but builds using the variable `TEST_SOURCES`, globbed in its parent directory.  Fixing this yields a whole bunch of include errors for `"test_helpers.hxx"` and `"test_main.hxx"`, which exist a directory above.  I went ahead and changed all these includes to `"../test_helpers.hxx"` etc. but haven't committed it — would you like a PR for that or just do it yourself?  This can also be fixed with `<test_helpers.hxx>` and an `INCLUDE_DIRECTORIES(../)` in the CMake script if you prefer.

**Comment by JadeMatrix** [comment]
I think the latest version (https://github.com/jtv/libpqxx/compare/master...JadeMatrix:inserter-tuples) is pretty much feature-complete, if you could give it another review.  I added a few unit tests, but some more may be appropriate.  Since there's a lot of changes, [here's the diff since the last time (possibly) you looked at it](https://github.com/JadeMatrix/libpqxx/compare/be79995...JadeMatrix:inserter-tuples) — note it also contains your changes in master that I merged in.

I added a meta-`pqxx::string_traits` for optional types, and it does streamline the code a bit.  It also means that containers (such as `pqxx::dynamic_params`) can seamlessly contain optional types (including pointers).  This brings up my biggest issue with `tablewriter`: for containers it does a string conversion then matches that against a null-string.  With `pqxx::string_traits` for optionals this is no longer necessary, and in fact can give false positives if the user specifies a null-string that happens to appear in their container as an actual value.

Unfortunately fixing it is probably a breaking change.  It would be possible to do type detection (doing *a lot* of that already) to detect when the user uses a container of optionals with `tablewriter`/`tablereader`, but this will possibly give unexpected results when they use a container of non-optionals.  Instead of erroring on null values libpqxx will silently insert a null-string.  While this is the behavior current users of `tablewriter`/`tablereader` would expect, it's inconsistent with how other type conversions work in libpqxx, both in master and especially in my branch.

Do you think it's a good idea to break the interface at this point?  If current users expect no null values, it reduces the amount of checks they have to perform in their code — the checks if not removed become redundant but should not affect how their code compiles or its correctness.  The issue is with users who *do* expect nulls.  Removing the null-string argument to the constructor would ideally cause their code to not compile and thus require attention; unfortunately there is a default argument for the null-string (an empty string).

For now, while I completely re-implemented how rows are created for `tablewriter` I was careful to replicate the original functionality.

**Comment by jtv** [comment]
I would love a CMake patch, thanks.  I saw your message about this in my email, then lost it, and thinking it was on a separate ticket somewhere, searched in vain for it!

**Comment by jtv** [comment]
Looking at the new diff.  I'm a little under the weather, but at least it gives me some time to sit quietly and stare at this.  :-)

Will this compile as C++11?  I thought things like the `template<typename>` template parameter, and the `auto` return type syntax require C++14 or C++17.  For the moment C++11 is still the baseline.

There's another possibility though: we freeze `tablereader` and `tablewriter` as they are, and present what you're doing as a replacement — which may require a newer C++ version, and that's fine.  That also gives you the freedom to change semantics without hurting existing code.

Next note: the two-parameter `field::get<>()` doesn't appeal to me.  I think I'd prefer just the single-parameter version, where the template argument will be some `optional<T>`.

My ulterior motive here is that I still dream of perhaps someday supporting binary data transfers.  Which may involve adding a template parameter in the same place, and I definitely don't want too many, especially with positional usage!

**Comment by jtv** [comment]
Oh, and I don't think I even noticed before, but as a word choice, "monobyte" is just genius.  Concise, clear, memorable.  Rolls off the tongue.  And since the "y" makes "byte" look like it might have come from Greek, the Greek "mono—" prefix works much better than the Latin "uni—" here.

**Comment by jtv** [comment]
From what you say, it sounds like the encoding stuff is pretty stable now.  I'd definitely make that a separate merge request, and get that in first...  Got to keep the size under control, or they turn into never-ending projects!
Also, I can't wait to start making use of it in code elsewhere.  :-)

**Comment by JadeMatrix** [comment]
Template template parameters have been around since before C++11, so no worries there.  The 2-parameter `get<>()` should be almost completely transparent from an API standpoint because the second parameter has a default value.  If either `std::optional<>` or `std::experimental::optional<>` are available, by default it acts like a single-parameter template returning one of those:

```cpp
  template<typename T, template<typename> class O
#if defined(PQXX_HAVE_OPTIONAL)
  = std::optional
#elif defined(PQXX_HAVE_EXP_OPTIONAL) && !defined(PQXX_HIDE_EXP_OPTIONAL)
  = std::experimental::optional
#endif
  > O<T> get() const
    ...
```

If neither are available then the user can *only* use it with a custom template type, whereas before the function simply didn't exist.  The second version with a single template parameter exists for custom non-templated optional-like types that may exist in legacy or niche codebases.

Of course this was mostly to maintain the API — I'm not too against changing it to the second one.  But then there's really not much difference between `get<>()` and `as<>()` — possibly merge them?

While I didn't really think about it too much, you're right that "uni-" also means one.  However in English it's mostly used to mean "as one," such as "united" or "Unicode" 😉  I'm now pondering what something called a "unibyte" could possibly be and how to use it…

I'll go ahead and make a PR for the CMake unit tests.  Extracting just the encoding stuff may be a bit difficult, because I only started on that several commits into the tuple work.  I can probably do a multi-part rebase, I'll need to check if any changes got mixed together in single commits.

**Comment by jtv** [comment]
(Note on counting bytes: the kilobyte-vs-kibibyte debate is nonsense.  A byte is _8.192 bits,_ which we normally round to 8.  A kilobyte has 24 _leap bytes._  It's not so complicated.)

You're right about `get`.  I'd forgotten that it's a distinct thing from `as` for this exact purpose.

**Comment by jtv** [comment]
I finally got to your `string_traits` implementation.  It's more generic and powerful than what I had in mind...  I was thinking `template<typename T> string_traits<std::optional<T>>`, and then re-doing that for each optional-like storage type.  But this'll cover them all in one fell swoop.

**Comment by JadeMatrix** [comment]
Well regarding `get<>()`/`as<>()`, I just realized that with the meta-`string_traits<>` for optionals they practically *are* the same thing anyways.  Both `get<int>()` and `as<optional<int>>()` return `optional<int>`.  With a slight tweak to the implementation of `as<>()` (to support non-default-constructable optionals) I should be able to write `get<>()` simply as:

```cpp
  template<
    typename T,
    template<typename> class O
#if defined(PQXX_HAVE_OPTIONAL)
      = std::optional
#elif defined(PQXX_HAVE_EXP_OPTIONAL) && !defined(PQXX_HIDE_EXP_OPTIONAL)
      = std::experimental::optional
#endif
  > using get = as<O<T>>;
```

**Comment by jtv** [comment]
I like that.  Once `std::optional` becomes the undisputed standard, `get()` will be the simpler thing to do, and we can probably just leave `as()` for people who like to do it the hard, old-fashioned way.

I'm not too bothered about `std::experimental::optional`.  I've had to throw it out in other cases where it created problems.  I hope we're close enough to C++17 as a baseline that a transitional solution isn't worth the effort.

**Comment by JadeMatrix** [comment]
It's a little late, but congrats on libpqxx's birthday!

I've been a bit preoccupied lately, but I think I have time to wrap this up now.  If you still want me to split the encoding stuff and `tablereader`/`tablewriter` stuff into separate branches I can probably do that, but it may be more straightforward to just include them in the next major or minor release, depending on your preference.  If you want to complete replace the existing `tablereader`/`tablewriter` I can clean up the interfaces a bit.  That may lead to more opportunities to use the type utils across libpqxx, which in turn strengthens the case for splitting off encodings support as something that's already *done*.  Let me know what you want to do.

Regarding moving up the target C++ version, I tend towards supporting the absolute lowest in which I can still reasonably accomplish desired functionality.  While `std::optional` is an incredibly useful tool for DB APIs (really wish they had included it and `std::void_t` in C++11), I've ensured that it isn't required for any of the type utils so there's no pressure there; in fact it may make it easier to integrate newer libpqxx into older codebases.  Just my 2¢.

**Comment by jtv** [comment]
I don't want to keep demanding more of your time before we can get this wonderful stuff merged.  So yes, _adding replacement classes_ for `tablereader` and `tablewriter` sounds better and better.  It means we can keep the old classes unchanged, and you get a lot more freedom to design things your way and think ahead to when `std::optional` is no longer a question.  And I'm OK with merging the encoding work as part of that in one go, really.

In which case, the next major decision is how to name the new classes.  The name `tablereader` was no longer a good fit in any case: `COPY TO stdout` can take any query, and that's a helpful thing to allow.  It just becomes hard to maintain the symmetry in the name.  I wouldn't want names like `istream` and `ostream` either, because they leave you wondering whether the "in" and "out" are relative to the databsae or to the C++ code.  So maybe `stream_to` and `stream_from`?

**Comment by JadeMatrix** [comment]
To match stdlib naming conventions my first thought would be to use `otablestream` and `itablestream`, or `orowstream`/`irowstream`.  Those aren't great to be honest.  `tablewriter2`/`tablereader2` is another option but probably worse.  Keeping the same name and using inline version namespaces is a C++11 approach, but that means doing that for the entire library.

How about `write_to`/`read_from`, or just `copy_to`/`copy_from`?

**Comment by JadeMatrix** [comment]
On second thought, the names `copy_to`/`copy_from` don't make sense as C++ constructs.  `COPY TO` copies data from a table to an output, and `COPY FROM` copies from an input to a table; this makes sense within a query:

```sql
COPY my.table TO STDOUT; -- Read data from table
COPY my.table FROM STDIN; -- Write data to table
```

If the C++ class names mirrored the underlying commands, it would read like this:

```cpp
auto ct = copy_to{ "my.table" }; // Read data from table
auto cf = copy_from{ "my.table" }; // Write data to table
```

As the position of the table name is lost in reading the constructor, this seems backwards.  "`copy_from` writes data?" someone might wonder.  So maybe mirroring the Postgres commands might not be the right choice.

I bring this up because I already mixed them up when moving them to separate headers 😉

**Comment by jtv** [comment]
Good point — I didn't see that coming!

So then how about `stream_from_table`, `stream_from_query`, and `stream_to_table`?  (That's adding hypothetical explicit support for copying from a query, which has been on my mind as well.)

**Comment by jtv** [comment]
Completely forgot that we had this issue still open.  I think we can give it a happy ending at this point.  :-)
